spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -368,19 +357,18 @@ private void doSendAndReceive(Message<?> message) {
 				}
 			}
 			finally {
-				if (KafkaInboundGateway.this.retryTemplate == null) {
-					ATTRIBUTES_HOLDER.remove();
-				}
+				ATTRIBUTES_HOLDER.remove();
 			}
 		}
 
 		private Message<?> enhanceHeadersAndSaveAttributes(Message<?> message, ConsumerRecord<K, V> record) {
 			Message<?> messageToReturn = message;
+			AttributeAccessor retryContext = ATTRIBUTES_HOLDER.get();
```

Comment:
We use the type `RetryContext` here instead of casting in other sections of the code. WDYT?

Reply:
We cannot do this because `ATTRIBUTES_HOLDER` not always holds exactly `RetryContext`.
Sometimes it is an `ErrorMessageUtils.getAttributeAccessor()`